Touch Bar Simulator/util.swift

```diff
@@ -53,6 +84,76 @@ extension NSView {
 	}
 }
 
+final class AssociatedObject<T: Any> {
+	subscript(index: Any) -> T? {
+		get {
+			return objc_getAssociatedObject(index, Unmanaged.passUnretained(self).toOpaque()) as! T?
+		} set {
+			objc_setAssociatedObject(index, Unmanaged.passUnretained(self).toOpaque(), newValue, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
+		}
+	}
+}
+
+@objc
+protocol TargetActionSender: AnyObject {
+	var target: AnyObject? { get set }
+	var action: Selector? { get set }
+}
+
+extension NSControl: TargetActionSender {}
+extension NSMenuItem: TargetActionSender {}
+
+private final class ActionClosureCaller: NSObject {
+	@objc
+	func callClosure(_ sender: TargetActionSender) {
+		onAction?(sender)
```

Comment:
The problem with this is that the `sender` will be typed as `TargetActionSender` and not the actual type, like `NSButton`. You need to use some generics here to get the type passed through.

Reply:
Not sure how possible this is, unfortunately.
* `@objc` methods cannot be added to protocol extensions.
* `@objc` methods cannot be generic.
* If I make `ActionClosureCaller` generic, then it seems okay, but then there's no way to declare the association key:
  * Protocol extensions cannot contain static stored properties—the TargetActionSender extension can't hold it.
  * Generic structs cannot contain static stored properties—TargetActionSenderAssociatedKeys cannot be made generic.

So I don't know how to pass the type statically through the target-action system. Guidance appreciated if you think there's a way.

Reply:
It's possible. This is from one of my closed-source apps:

```swift
private var ControlActionClosureProtocolAssociatedObjectKey: UInt8 = 0

protocol ControlActionClosureProtocol: NSObjectProtocol {
	var target: AnyObject? { get set }
	var action: Selector? { get set }
}

private final class ActionTrampoline<T>: NSObject {
	let action: (T) -> Void

	init(action: @escaping (T) -> Void) {
		self.action = action
	}

	@objc
	func action(sender: NSControl) {
		action(sender as! T)
	}
}

extension ControlActionClosureProtocol {
	func onAction(_ action: @escaping (Self) -> Void) {
		let trampoline = ActionTrampoline(action: action)
		self.target = trampoline
		self.action = #selector(ActionTrampoline<Self>.action(sender:))
		objc_setAssociatedObject(self, &ControlActionClosureProtocolAssociatedObjectKey, trampoline, .OBJC_ASSOCIATION_RETAIN)
	}
}

extension NSControl: ControlActionClosureProtocol {}
extension NSMenuItem: ControlActionClosureProtocol {}
extension NSGestureRecognizer: ControlActionClosureProtocol {}
```

What I would change today, it to use a `.onAction` getter/setter instead of function, like you've done, as it makes it clear that only one listener can be added. It also makes it possible to replace the listener and read the listener. So would be great if you could adapt the above to use a getter/setter instead of function. And the `objc_setAssociatedObject` could be replaced with `AssociatedObject` class.